### PR TITLE
feat: add support for aws sdk v3 lambda

### PR DESCRIPTION
**Pull request by aryamohanan**
refs 136533

### CODE REVIEW

**Review by kirrg001 — COMMENTED** [review]
Great work. Added some comments :)

File: packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/index.js [hunk 1]
```diff
@@ -79,8 +80,10 @@ function shimSmithySend(originalSend) {
     const command = smithySendArgs[0];
 
     const serviceId = self.config && self.config.serviceId;
-    let awsProduct = serviceId && awsProducts.find(aws => aws.spanName === serviceId.toLowerCase());
-
+    let awsProduct =
+      serviceId &&
+      (awsProducts.find(aws => aws.spanName === serviceId.toLowerCase()) ||
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/index.js]: Done

File: packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/utils.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        const isCreationComplete = await isFunctionCreationComplete(functionName);
+        if (isCreationComplete) {
+          clearInterval(intervalId);
+          resolve(true);
+        }
+      } catch (error) {
+        clearInterval(intervalId);
+        resolve(false);
+      }
+    }, 100);
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/utils.js]: Done.

File: packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/app.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+const lambda = new LambdaClient(clientOpts);
+const lambdaV2 = new Lambda(clientOpts);
+const operations = {
+  invoke: {
+    FunctionName: functionName,
+    // "Event"|"RequestResponse"|"DryRun"
+    InvocationType: 'RequestResponse',
+    Payload: '{"ok": true}'
+  },
+  invokeAsync: {
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/app.js:33]: Removed invokeAsync when running tests with localstack, also added comments

File: packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    return cls.ns.runAndReturn(() => {
+      const span = cls.startSpan(SPAN_NAME, EXIT);
+      span.ts = Date.now();
+      span.stack = tracingUtil.getStackTrace(this.instrumentedMakeRequest, 1);
+      span.data[this.spanName] = this.buildSpanData(smithySendArgs[0], smithySendArgs[0].input);
+
+      this.propagateInstanaHeaders(smithySendArgs, span);
+      const callback = typeof smithySendArgs[1] === 'function' ? smithySendArgs[1] : smithySendArgs[2];
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js]: Done.

File: packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/instana_aws_product.js [hunk 5]
```diff
@@ -9,10 +9,12 @@ class InstanaAWSProduct {
   /**
    * @param {string} spanName
    * @param {Array.<string>} operations
+   * @param {string} customProductName
    */
-  constructor(spanName, operations) {
+  constructor(spanName, operations, customProductName) {
     this.spanName = spanName;
     this.operations = operations || [];
+    this.customProductName = customProductName || '';
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/instana_aws_product.js]: Updated to 'serviceName'.

File: packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/test_definition.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+const globalAgent = require('../../../../../globalAgent');
+const { verifyHttpRootEntry, verifyExitSpan } = require('@instana/core/test/test_util/common_verifications');
+const { isLocalStackDisabled } = require('./utils');
+const SPAN_NAME = 'aws.lambda.invoke';
+const functionName = 'wrapped-async';
+let appControls;
+
+const availableCtx = [null, '{"Custom": {"awesome_company": "Instana"}}', '{"Custom": "Something"}'];
+const requestMethods = ['async', 'promise', 'cb', 'promise-v2', 'cb-v2'];
+const availableOperations = ['invoke', 'invokeAsync'];
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/test_definition.js]: Added case for `GetFunctionCommand`

File: packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/app.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+ */
+
+'use strict';
+
+require('../../../../../../src')();
+
+const { LambdaClient, InvokeCommand, Lambda, InvokeAsyncCommand } = require('@aws-sdk/client-lambda');
+const express = require('express');
+const logPrefix = `AWS SDK v3 Lambda (${process.pid}):\t`;
+const functionName = process.env.AWS_LAMBDA_FUNCTION_NAME || 'wrapped-async';
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/collector/test/tracing/cloud/aws-sdk/v3/lambda/app.js]: Done.

File: packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+          });
+
+        return request;
+      }
+    });
+  }
+
+  buildSpanData(_operation, params) {
+    const spanData = {
+      function: (params && params.FunctionName) || ''
```
[the remark this answers is not in the document]
Comment (aryamohanan) [reply packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js]: Removed this assignement

File: packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        smithySendArgs[callbackIndex] = cls.ns.bind(function (err) {
+          self.finishSpan(err, span);
+          return originalCallback.apply(this, arguments);
+        });
+        return originalSend.apply(ctx, smithySendArgs);
+      } else {
+        const request = originalSend.apply(ctx, smithySendArgs);
+        request
+          .then(() => {
+                        this.finishSpan(null, span);
```
Comment (kirrg001) [inline packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js:95]: This looks like a wrong indention. Wondering how that is possible 🤔

**Review by kirrg001 — APPROVED** [review]
🥳

File: packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js (hunk 9 above)
Comment (aryamohanan) [reply packages/core/src/tracing/instrumentation/cloud/aws-sdk/v3/lambda.js]: Ah right, Thanks for pointing this, I will update this

**Pull request merged**